검색창 컴포넌트 및 포스트 리스트 페이지 구현

jma1020 opened the pull request:
# Pull Request
 - [x] Feature
 
 # Summary
- 검색창 구현
- posts index 페이지 구현


# Describe your changes
- React Portal을 이용한 검색창 구현 
- 기존 포스트 리스트와 검색결과에 따른 search api 구현
- 포스트 리스트 페이지 구현
- 포스트 리스트 서버사이드-렌더링 적용
- React query(useInfinityQuery)로 무한스크롤 적용


# Screenshots
### 검색창 UI
![image](https://user-images.githubusercontent.com/45344418/198690673-db018449-ee72-4a3a-b149-db02a49c72a1.png)

### 게시물 리스트 UI 
![image](https://user-images.githubusercontent.com/45344418/198690891-a6368564-6a3d-44d3-b31d-fe70b744f079.png)

![image](https://user-images.githubusercontent.com/45344418/198690983-8c062324-802d-4abc-9e6d-49c93d7a29ef.png)

# ETC
- 현재 PostCategory버튼에 색상과 검색창 enter key event 고민 중
- 추가로 정렬 드롭다운 추가
- 조건에 맞는 게시물 없을 시 보여줄 컴포넌트 추가

jma1020 committed:
feat: React Portal 세팅

jma1020 committed:
feat: SearchBar 검색창 컴포넌트 생성 및 GNB 레이아웃에 적용

jma1020 committed:
refact: searchBar state -> useToggle사용으로 변경

jma1020 committed:
refact: styled 파일 분리

jma1020 committed:
refact: Hydration 오류 방지를 위해 Portal 컴포넌트 useState, useEffect
사용

jma1020 committed:
feat: SearchBar UI 수정

- SearchType 으로 인한 UI 수정

jma1020 committed:
feat: search api 생성

jma1020 committed:
feat: Search Bar 검색 버튼 Link 연결

jma1020 committed:
feat: 카테고리 버튼 컴포넌트 생성

jma1020 committed:
fix: PostList 컴포넌트 posts data 옵셔널체인닝 수정

jma1020 committed:
feat: posts 페이지 구현

- serverside 렌더링
- react query 사용

jma1020 committed:
refact: useInfinityQuery 형식에 맞도록 api 수정

jma1020 committed:
refact: Category 컴포넌트 post-category 상수 파일을 사용

- post-category 상수 파일 index signature 추가

jma1020 committed:
feat: PostList 페이지 useInfinityQuery 적용

- PostList컴포넌트 useInfinityQuery response 형식에 맞게 변경
- PostPage useInfinityQuery 적용
- getserverSideProps Hydration serializing error 해결

jma1020 committed:
Merge branch 'develop' of https://github.com/TRIPONG-2022/frontend into feature/search-bar

jma1020 committed:
refact: CategoryButton 컴포넌트 네이밍 변경

jma1020 committed:
feat: 정렬 params 추가

jma1020 committed:
feat: 무한 스크롤 적용 및 간단한 style, 네이밍 수정

jma1020 committed:
fix: 폴더 네이밍 변경

jma1020 committed:
feat: PostListNotFound 컴포넌트 생성

jma1020 committed:
feat: PostListNotFound 컴포넌트 사용

jma1020 committed:
refact: PostListQuery hooks 분리

sangrimlee commented on api/search.ts:
```diff
@@ -0,0 +1,63 @@
+    params: { searchType, keyword },
+  });
+
+  console.log(data);
```
console.log 발견 ⚠️

jma1020 replied:
수정했습니다ㅎㅎ

sangrimlee commented on api/search.ts:
```diff
@@ -0,0 +1,63 @@
+  pageParam?: number,
+) => {
+  if (params.keyword) {
+    const { data } = await axios.get(
+      `${process.env.NEXT_PUBLIC_BASE_URL}/search`,
```
SSR때문에 instance를 쓰지 않은 건가요?
그렇다면, Public한 자원에 접근하는 isntance랑 Private한 자원에 접근하는 instance를 분리해도 좋을 것 같아요.

jma1020 replied:
생각도 못했는데, 감사합니다 ㅎㅎㅎ
기존 instance 파일에 생성하지 않고 public-instance 파일을 추가로 만들어서 올려두었습니다!

sangrimlee commented on components/PostCategory/PostCategoryButton.tsx:
```diff
@@ -0,0 +1,37 @@
+        active={postCategory === ''}
+        onClick={() => setPostCategory('')}
+      >
+        전체 결과
```
카테고리니깐 전체 결과 보다는 전체라던지 다른 사이트 참고해서 조금 자연스러운 Label 찾아보면 좋을 것 같아요.

jma1020 replied:
![image](https://user-images.githubusercontent.com/45344418/198887905-327568b2-b570-4e1e-abd6-4425f650e268.png)
커뮤니티 사이트 참고해보고서 전체를 넣기로 결정했습니다!!.. 😀

sangrimlee commented on components/post/PostList/PostList.tsx:
```diff
@@ -18,13 +18,16 @@ interface PostListProps {
  * - lg: Mobile 1, Tablet 2, Desktop 3
  */
 export default function PostList({ posts, size = 'md' }: PostListProps) {
+  console.log(posts);
```
console.log 발견 ⚠️

sangrimlee commented on components/post/PostList/PostList.tsx:
```diff
@@ -6,7 +6,7 @@ import PostItem from '../PostItem';
 
 interface PostListProps {
-  posts: Post[];
+  posts: { pages: Post[][] } | undefined;
```
Props를 바꾸지말고, 바뀌어진 Props를 내려보내면 좋지 않을까요?
여기서는, Optional(undefined)인 경우만 추가하고, pages안에 2차원 배열 post를 재가공하거나, `Post[][]`으로 변경하는 것도 괜찮아보입니다!

jma1020 replied:
이 부분 수정해서 올려두었습니다!!😄

sangrimlee commented on components/searchbar/SearchBar.styled.tsx:
```diff
@@ -0,0 +1,69 @@
+  left: 0;
+  right: 0;
+  height: ${HEADER_HEIGHT};
+  z-index: 9999;
```
주어진 z-index 상수에서 쓰면 좋을 것 같아요!

jma1020 replied:
수정했습니다 ㅎㅎ

sangrimlee commented on components/searchbar/SearchBar.tsx:
```diff
@@ -0,0 +1,91 @@
+  const router = useRouter();
+
+  const [searchInput, setSearchInput] = useState('');
+  const [searchType, setSearchType] = useState<string>('TITLE');
```
SearchType에 대한 타입을 enum이라던지, union 타입을 활용해서 표현하면 더 좋을 것 같아요!

jma1020 replied:
enum으로 처리했습니다ㅎㅎ  피드백 감사합니다😊

sangrimlee commented on components/searchbar/SearchBar.tsx:
```diff
@@ -0,0 +1,91 @@
+  const [searchInput, setSearchInput] = useState('');
+  const [searchType, setSearchType] = useState<string>('TITLE');
+
+  const setState =
+    (setSearchParams: React.Dispatch<React.SetStateAction<string>>) =>
+    (value: string) =>
+      setSearchParams((prev) => value);
```
`onChangeOption: (value:T) => void;`이거니깐, onChangeSearchType = (value: SearchType) => void; 형식의 함수로만 사용하면 충분하지 않을까요?
굳이, 고차함수로 표현안해도 될 것 같아요.

jma1020 replied:
감사합니다!!  수정하면서 이해하게 되었네요 ㅎㅎ 😅

sangrimlee commented on components/searchbar/SearchBar.tsx:
```diff
@@ -0,0 +1,91 @@
+          </Styled.SelectWrapper>
+          <Styled.SearchInputWrapper>
+            <Styled.SearchInput
+              onChange={(e: any) => {
```
`any`타입은 왠만하면 사용하지 않는 게 좋아요!
`React.ChangeEvent<HTMLInputElement>`

jma1020 replied:
수정했습니다 😊

sangrimlee commented on components/searchbar/SearchBar.tsx:
```diff
@@ -0,0 +1,91 @@
+                icon="SearchIcon"
+                title="검색아이콘"
+                aria-label="검색아이콘"
+                onClick={() =>
+                  router.push(
+                    `/posts?searchType=${searchType}&keyword=${searchInput}`,
+                  )
```
함수로 빼는 것도 고민해볼법하다고 보입니다! 왜나하면, 저 URL이 어디로 가는지 명시적으로 나타낼 수 있을 것 같아요!

jma1020 replied:
피드백 감사합니다!!  반영했습니다 😀

sangrimlee commented on components/shared/Portal/Portal.tsx:
```diff
@@ -0,0 +1,23 @@
+    }
+  }, [selector]);
+
+  if (!element) return <></>;
```
이런 경우에 그냥 null을 Return해도 됩니다. 굳이 Fragment를 Return 하지 않아도 됩니다.
React.ReactNode의 타입을 한 번 찾아가보시면 도움이 될 것 같아요!

jma1020 replied:
반영했습니다 ㅎㅎ

sangrimlee commented on hooks/usePostListQuery.ts:
```diff
@@ -0,0 +1,37 @@
+      },
+
+      select: (data) => ({
+        pages: [...data.pages].map((list) =>
+          postCategory
+            ? list.filter(
+                ({ category }: { category: string }) =>
+                  category === postCategory,
+              )
+            : list,
```
Post 타입의 category가 PostCategory라는 enum형식이기 때문에 위에서 props로 받는 타입부터 차근차근 수정해나가면 좋을 것 같아요!

jma1020 committed:
fix: console.log 삭제

jma1020 committed:
refact: SearchBar 컴포넌트 Link router 함수 생성 및 SearchType 타입 생성

jma1020 committed:
fix: 간단한 return 수정

jma1020 committed:
feat: axios -> public-instance 생성 및 적용

- SSR에서 주로 사용될 것을 고려해서 생성

jma1020 committed:
fix: z-index 컨벤션에 따른 변경

jma1020 committed:
refact: 폴더 및 파일 네이밍 변경

- Button -> Tap으로 수정

jma1020 committed:
fix: portal div 렌더링 위치 변경

- Main 밑으로 이동

jma1020 committed:
fix: 임시 Z_INDEX 값 OFF_CANVAS 설정 (GNB Z_INDEX 값으로 인한)

- GNB 에 Z_INDEX HEADER에 값이 FIXED 보다 높아 임시로 설정

jma1020 committed:
refact: 기존에 불필요하던 고차함수 형태 수정

jma1020 committed:
fix: 불필요한 주석 삭제

jma1020 committed:
refact: postList 컴포넌트 props data 형식 변경, 임시 정렬 버튼 생성

jma1020 committed:
fix: z-index 상수 수정 및 적용

jma1020 committed:
fix: entries type 이슈로 일시적인 타입 단언 적용

jma1020 committed:
refact: PostCategory  타입 적용

jma1020 committed:
refact: PostList 컴포넌트 props 변경에 따른 타입 수정

jma1020 committed:
feat: searchSlice 생성

jma1020 committed:
refact: 기존 isSearch toggle state -> redux search 로 변경

jma1020 committed:
fix: console 삭제

jma1020 committed:
feat: Search UI 추가로 인한 MainLayout 수정

jma1020 committed:
refact: 사용하지 않는 코드 삭제 및 스타일 수정

jma1020 committed:
refact: Search UI 리팩토링

jma1020 committed:
feat: Search state 핸들링 추가

jma1020 committed:
refact: 기존 PostList props 형식에 맞도록 query data 가공

jma1020 committed:
fix: 간단한 스타일 수정

jma1020 committed:
feature: Tab 컴포넌트 생성

jma1020 committed:
fix: api 에러로 인한 간단한 임시 수정

jma1020 committed:
fix: 불필요한 코드 삭제

jma1020 committed:
refact: search api 에러 방지 추가

jma1020 committed:
fix: 불필요한 코드 삭제

jma1020 committed:
refact: MainLayout 컴포넌트 리팩토링

jma1020 committed:
feat: api 정렬 기능 사용

jma1020 committed:
refact: api 정렬 기능 추가에 따른 posts 페이지 수정

jma1020 committed:
feat: PostCategoryContent 분리 및 scroll 에 따른 sticky 동작 추가

- sticky 형식으로 스타일 변경
- scroll Up을 파악할 수 있는 커스텀훅 생성
- 모바일 환경에서 카테고리 클릭 시 스크롤 이동 이벤트 추가

jma1020 committed:
refact: useScroll hook 리팩토링

- 범용성을 위해 scrollDirection, scrollY util 추가

jma1020 committed:
refact: PostCategory 리팩토링

- 변수 네이밍 리팩토링
- useScroll 변경에 따른 리팩토링

jma1020 committed:
feat: requestAnimationFrame 함수 util 생성

sangrimlee commented on api/auth.ts:
```diff
@@ -40,7 +39,7 @@ export const login = async ({ loginId, password }: LoginType) => {
   try {
     const data = await instance.get<ConfirmUserResponse>('/users/profile');
-
+    console.log(data);
```
⚠️ console.log 발견

jma1020 replied:
엇 console 수정을 또 엄청 많았네요 ㅠㅠ... 수정했습니다!!😀 감사합니다~~

sangrimlee commented on api/auth.ts:
```diff
@@ -57,6 +56,7 @@ export const userConfirm = async () => {
     };
   } catch (err) {
     if (axios.isAxiosError(err)) {
+      console.log(err);
```
⚠️ console.log 발견

jma1020 replied:
수정했습니다!!😀 감사합니다~~

sangrimlee commented on api/search.ts:
```diff
@@ -0,0 +1,71 @@
+
+    return data;
+  } catch (err) {
+    console.log(err);
```
⚠️ console.log 발견

jma1020 replied:
수정했습니다!!😀 감사합니다

sangrimlee commented on components/PostCategoryTap/PostCategoryContent.tsx:
```diff
@@ -0,0 +1,45 @@
+      ref={ref}
+      active={active}
+      onClick={() => {
+        setPostCategory(keyValue as PostCategory);
+
+        tapScrollX.current = ref.current?.getBoundingClientRect().x;
+        containerRef.current?.scrollTo({
+          behavior: 'smooth',
+          left: containerRef.current?.scrollLeft + tapScrollX.current! - 20,
+        });
```
단순하게 버튼이 클릭되었을 때, 스크롤을 위해서 너무 많은 Props를 넘겨 받는 것 같아요.
스크롤 동작을 위한 함수만을 Props로 받는 것은 어떨까요?

jma1020 replied:
handleTapScrollX 함수를 새로 생성해서 사용하도록 변경해주었습니다 ㅎㅎ

jma1020 committed:
refact: Scroll Event 리팩토링

- usePrevious hook 생성
- RequestAnimationFrame hook 분

sangrimlee commented on components/PostCategoryTap/PostCategoryTap.tsx:
```diff
@@ -0,0 +1,61 @@
+  return (
+    <Styled.PostCategoryTapContiner
+      ref={containerRef}
+      visibleOnScroll={scrollY > 160 && scrollDirection === 'UP'}
```
`scrollY > 160 && scrollDirection === 'UP'`이 부분을 변수로 빼내는 게 좋을 것 같아요! 정확하게 코드만 봤을 때 어떻게 동작하는지 예상하기 힘든 것 같아요.

jma1020 replied:
감사합니다ㅎㅎ useMemo 변수로 변경해주었습니다!!😊

jma1020 committed:
refact: props로 내려받는 코드 리팩토링 (피드백 적용)

jma1020 committed:
refact: 사용하지않는 console 삭제

seongjin2427 commented on api/search.ts:
```diff
@@ -0,0 +1,68 @@
+  keyword,
+}: {
+  searchType: string | string[] | undefined;
+  keyword: string | string[] | undefined;
```
searchType?: , keyword?: 로 더 간단하게 표현 해줄 수 있을 것 같습니당! ㅎㅎ

jma1020 replied:
수정했습니다 ㅎㅎ😊

jma1020 committed:
refact: 파일 디렉터리 변경

seongjin2427 commented on utils/scroll.ts:
```diff
@@ -0,0 +1,8 @@
+  }
+
+  return { rAFId, rAFFunc };
+}
```
헛 혹시 createRAF가 사용되는 곳이 있나요!
없다면 지워도 괜찮지 않을까 생각합니당 하핳

jma1020 replied:
감사합니다 ㅎㅎ 삭제했습니다!!

seongjin2427 reviewed (approved):
고생하셨습니다!!

jma1020 committed:
fix: 피드백 적용

jma1020 committed:
Merge branch 'develop' into feature/search-bar
